File: sdk/python/term_sdk/llm.py
"""
LLM Client for Term Challenge agents.

Supports multiple providers:
- OpenRouter (default): anthropic/claude-3-haiku, openai/gpt-4o, etc.
- OpenAI: gpt-4o, gpt-4o-mini, gpt-3.5-turbo
- Anthropic: claude-3-opus, claude-3-sonnet, claude-3-haiku

Example:
    ```python
    from term_sdk import LLM
    
    # OpenRouter (default)
    llm = LLM(model="anthropic/claude-3-haiku")
    
    # Direct OpenAI
    llm = LLM(provider="openai", model="gpt-4o")
    
    # Direct Anthropic
    llm = LLM(provider="anthropic", model="claude-3-haiku-20240307")
    
    # Ask a question
    response = llm.ask("What is 2+2?")
    print(response.text)
    
    # Chat with messages
    response = llm.chat([
        {"role": "system", "content": "You are helpful."},
        {"role": "user", "content": "Hello!"}
    ])
    ```
"""

import os
import sys
import json
import time
from dataclasses import dataclass
from typing import Optional, List, Dict, Any
import httpx


@dataclass
class LLMResponse:
    """Response from LLM."""
    text: str
    model: str
    tokens: int = 0
    cost: float = 0.0
    latency_ms: int = 0
    
    def json(self) -> Optional[Dict]:
        """Parse response as JSON."""
        try:
            text = self.text.strip()
            # Remove markdown code blocks
            if "```" in text:
                import re
                match = re.search(r'```(?:json)?\s*(\{.*?\})\s*```', text, re.DOTALL)
                if match:
                    text = match.group(1)
            start = text.find('{')
            end = text.rfind('}')
            if start >= 0 and end > start:
                return json.loads(text[start:end + 1])
        except:
            pass
        return None


# Provider configurations
PROVIDERS = {
    "openrouter": {
        "url": "https://openrouter.ai/api/v1/chat/completions",
        "key_env": "OPENROUTER_API_KEY",
        "headers": lambda key: {
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
        }
    },
    "openai": {
        "url": "https://api.openai.com/v1/chat/completions",
        "key_env": "OPENAI_API_KEY",
        "headers": lambda key: {
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
        }
    },
    "anthropic": {
        "url": "https://api.anthropic.com/v1/messages",
        "key_env": "ANTHROPIC_API_KEY",
        "headers": lambda key: {
            "x-api-key": key,
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01",
        }
    },
}

# Model pricing (per 1M tokens) - input/output
PRICING = {
    # OpenRouter models
    "anthropic/claude-3-haiku": (0.25, 1.25),
    "anthropic/claude-3-sonnet": (3.0, 15.0),
    "anthropic/claude-3-opus": (15.0, 75.0),
    "anthropic/claude-3.5-sonnet": (3.0, 15.0),
    "openai/gpt-4o": (5.0, 15.0),
    "openai/gpt-4o-mini": (0.15, 0.6),
    "openai/gpt-3.5-turbo": (0.5, 1.5),
    "google/gemini-pro": (0.5, 1.5),
    "meta-llama/llama-3-70b": (0.8, 0.8),
    # Direct OpenAI
    "gpt-4o": (5.0, 15.0),
    "gpt-4o-mini": (0.15, 0.6),
    "gpt-3.5-turbo": (0.5, 1.5),
    # Direct Anthropic
    "claude-3-haiku-20240307": (0.25, 1.25),
    "claude-3-sonnet-20240229": (3.0, 15.0),
    "claude-3-opus-20240229": (15.0, 75.0),
}


def _log(msg: str):
    print(f"[llm] {msg}", file=sys.stderr)


class LLM:
    """
    LLM client for multiple providers.
    
    Args:
        provider: "openrouter" (default), "openai", or "anthropic"
        model: Model name (e.g., "anthropic/claude-3-haiku")
        api_key: API key (or set via environment variable)
        temperature: Sampling temperature (0.0 - 2.0)
        max_tokens: Maximum response tokens
        timeout: Request timeout in seconds
    
    Example:
        ```python
        llm = LLM(model="anthropic/claude-3-haiku")
        
        # Simple question
        response = llm.ask("What is Python?")
        print(response.text)
        
        # With system prompt
        response = llm.ask(
            "Write hello world",
            system="You are a Python expert."
        )
        
        # Chat with history
        response = llm.chat([
            {"role": "system", "content": "You are helpful."},
            {"role": "user", "content": "Hello!"},
            {"role": "assistant", "content": "Hi there!"},
            {"role": "user", "content": "How are you?"}
        ])
        ```
    """
    
    def __init__(
        self,
        provider: str = "openrouter",
        model: str = "anthropic/claude-3-haiku",
        api_key: Optional[str] = None,
        temperature: float = 0.3,
        max_tokens: int = 1024,
        timeout: int = 60,
    ):
        self.provider = provider
        self.model = model
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.timeout = timeout
        
        # Get provider config
        if provider not in PROVIDERS:
            raise ValueError(f"Unknown provider: {provider}")
        self._config = PROVIDERS[provider]
        
        # Get API key
        self.api_key = api_key or os.environ.get(self._config["key_env"], "")
        if not self.api_key:
            _log(f"Warning: {self._config['key_env']} not set")
        
        # Stats
        self.total_tokens = 0
        self.total_cost = 0.0
        self.request_count = 0
        
        # HTTP client
        self._client = httpx.Client(timeout=timeout)
    
    def ask(
        self,
        prompt: str,
        system: Optional[str] = None,
        **kwargs
    ) -> LLMResponse:
        """
        Ask a simple question.
        
        Args:
            prompt: User prompt
            system: Optional system prompt
            **kwargs: Override model, temperature, max_tokens
        
        Returns:
            LLMResponse with text, tokens, cost
        """
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})
        return self.chat(messages, **kwargs)
    
    def chat(
        self,
        messages: List[Dict[str, str]],
        **kwargs
    ) -> LLMResponse:
        """
        Chat with message history.
        
        Args:
            messages: List of {"role": "user/assistant/system", "content": "..."}
            **kwargs: Override model, temperature, max_tokens
        
        Returns:
            LLMResponse with text, tokens, cost
        """
        model = kwargs.get("model", self.model)
        temperature = kwargs.get("temperature", self.temperature)
        max_tokens = kwargs.get("max_tokens", self.max_tokens)
        
        start = time.time()
        
        try:
            if self.provider == "anthropic":
                response = self._chat_anthropic(messages, model, temperature, max_tokens)
            else:
                response = self._chat_openai(messages, model, temperature, max_tokens)
            
            response.latency_ms = int((time.time() - start) * 1000)
            
            # Update stats
            self.total_tokens += response.tokens
            self.total_cost += response.cost
            self.request_count += 1
            
            _log(f"{model}: {response.tokens} tokens, ${response.cost:.4f}, {response.latency_ms}ms")
            return response
            
        except Exception as e:
            _log(f"Error: {e}")
            raise
    
    def _chat_openai(
        self,
        messages: List[Dict[str, str]],
        model: str,
        temperature: float,
        max_tokens: int,
    ) -> LLMResponse:
        """OpenAI/OpenRouter compatible API."""
        headers = self._config["headers"](self.api_key)
        
        payload = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        
        response = self._client.post(
            self._config["url"],
            headers=headers,
            json=payload,
        )
        response.raise_for_status()
        data = response.json()
        
        text = data["choices"][0]["message"]["content"]
        usage = data.get("usage", {})
        prompt_tokens = usage.get("prompt_tokens", 0)
        completion_tokens = usage.get("completion_tokens", 0)
        total_tokens = prompt_tokens + completion_tokens
        
        # Calculate cost
        pricing = PRICING.get(model, (0.5, 1.5))
        cost = (prompt_tokens * pricing[0] + completion_tokens * pricing[1]) / 1_000_000
        
        return LLMResponse(
            text=text,
            model=model,
            tokens=total_tokens,
            cost=cost,
        )
    
    def _chat_anthropic(
        self,
        messages: List[Dict[str, str]],
        model: str,
        temperature: float,
        max_tokens: int,
    ) -> LLMResponse:
        """Anthropic native API."""
        headers = self._config["headers"](self.api_key)
        
        # Extract system message
        system = None
        user_messages = []
        for msg in messages:
            if msg["role"] == "system":
                system = msg["content"]
            else:
                user_messages.append(msg)
        
        payload = {
            "model": model,
            "messages": user_messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        if system:
            payload["system"] = system
        
        response = self._client.post(
            self._config["url"],
            headers=headers,
            json=payload,
        )
        response.raise_for_status()
        data = response.json()
        
        text = data["content"][0]["text"]
        usage = data.get("usage", {})
        prompt_tokens = usage.get("input_tokens", 0)
        completion_tokens = usage.get("output_tokens", 0)
        total_tokens = prompt_tokens + completion_tokens
        
        # Calculate cost
        pricing = PRICING.get(model, (0.25, 1.25))
        cost = (prompt_tokens * pricing[0] + completion_tokens * pricing[1]) / 1_000_000
        
        return LLMResponse(
            text=text,
            model=model,
            tokens=total_tokens,
            cost=cost,
        )
    
    def close(self):
        """Close HTTP client."""
        self._client.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, *args):
        self.close()

File: sdk/rust/src/llm.rs
//! LLM client for Term Challenge agents.
//!
//! Supports multiple providers:
//! - OpenRouter (default)
//! - OpenAI
//! - Anthropic

use std::env;
use std::time::Instant;
use serde::{Deserialize, Serialize};

/// LLM provider.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Provider {
    OpenRouter,
    OpenAI,
    Anthropic,
}

impl Provider {
    fn url(&self) -> &'static str {
        match self {
            Provider::OpenRouter => "https://openrouter.ai/api/v1/chat/completions",
            Provider::OpenAI => "https://api.openai.com/v1/chat/completions",
            Provider::Anthropic => "https://api.anthropic.com/v1/messages",
        }
    }
    
    fn env_key(&self) -> &'static str {
        match self {
            Provider::OpenRouter => "OPENROUTER_API_KEY",
            Provider::OpenAI => "OPENAI_API_KEY",
            Provider::Anthropic => "ANTHROPIC_API_KEY",
        }
    }
}

/// LLM response.
#[derive(Debug, Clone)]
pub struct LLMResponse {
    /// Response text
    pub text: String,
    /// Model used
    pub model: String,
    /// Total tokens used
    pub tokens: u32,
    /// Cost in USD
    pub cost: f64,
    /// Latency in milliseconds
    pub latency_ms: u64,
}

/// Chat message.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub role: String,
    pub content: String,
}

impl Message {
    pub fn system(content: impl Into<String>) -> Self {
        Self { role: "system".to_string(), content: content.into() }
    }
    
    pub fn user(content: impl Into<String>) -> Self {
        Self { role: "user".to_string(), content: content.into() }
    }
    
    pub fn assistant(content: impl Into<String>) -> Self {
        Self { role: "assistant".to_string(), content: content.into() }
    }
}

/// LLM client.
pub struct LLM {
    provider: Provider,
    model: String,
    api_key: String,
    temperature: f32,
    max_tokens: u32,
    client: reqwest::blocking::Client,
    
    // Stats
    pub total_tokens: u32,
    pub total_cost: f64,
    pub request_count: u32,
}

impl LLM {
    /// Create new LLM client with OpenRouter.
    pub fn new(model: impl Into<String>) -> Self {
        Self::with_provider(Provider::OpenRouter, model)
    }
    
    /// Create new LLM client with specific provider.
    pub fn with_provider(provider: Provider, model: impl Into<String>) -> Self {
        let api_key = env::var(provider.env_key()).unwrap_or_default();
        if api_key.is_empty() {
            eprintln!("[llm] Warning: {} not set", provider.env_key());
        }
        
        Self {
            provider,
            model: model.into(),
            api_key,
            temperature: 0.3,
            max_tokens: 1024,
            client: reqwest::blocking::Client::new(),
            total_tokens: 0,
            total_cost: 0.0,
            request_count: 0,
        }
    }
    
    /// Set temperature.
    pub fn temperature(mut self, t: f32) -> Self {
        self.temperature = t;
        self
    }
    
    /// Set max tokens.
    pub fn max_tokens(mut self, t: u32) -> Self {
        self.max_tokens = t;
        self
    }
    
    /// Set API key.
    pub fn api_key(mut self, key: impl Into<String>) -> Self {
        self.api_key = key.into();
        self
    }
    
    /// Ask a simple question.
    pub fn ask(&mut self, prompt: &str) -> Result<LLMResponse, String> {
        self.chat(&[Message::user(prompt)])
    }
    
    /// Ask with system prompt.
    pub fn ask_with_system(&mut self, system: &str, prompt: &str) -> Result<LLMResponse, String> {
        self.chat(&[Message::system(system), Message::user(prompt)])
    }
    
    /// Chat with messages.
    pub fn chat(&mut self, messages: &[Message]) -> Result<LLMResponse, String> {
        let start = Instant::now();
        
        let response = if self.provider == Provider::Anthropic {
            self.chat_anthropic(messages)?
        } else {
            self.chat_openai(messages)?
        };
        
        let mut response = response;
        response.latency_ms = start.elapsed().as_millis() as u64;
        
        self.total_tokens += response.tokens;
        self.total_cost += response.cost;
        self.request_count += 1;
        
        eprintln!("[llm] {}: {} tokens, ${:.4}, {}ms", 
            response.model, response.tokens, response.cost, response.latency_ms);
        
        Ok(response)
    }
    
    fn chat_openai(&self, messages: &[Message]) -> Result<LLMResponse, String> {
        #[derive(Serialize)]
        struct OpenAIRequest<'a> {
            model: &'a str,
            messages: &'a [Message],
            temperature: f32,
            max_tokens: u32,
        }
        
        #[derive(Deserialize)]
        struct OpenAIResponse {
            choices: Vec<Choice>,
            usage: Option<Usage>,
        }
        
        #[derive(Deserialize)]
        struct Choice {
            message: ChoiceMessage,
        }
        
        #[derive(Deserialize)]
        struct ChoiceMessage {
            content: String,
        }
        
        #[derive(Deserialize)]
        struct Usage {
            prompt_tokens: u32,
            completion_tokens: u32,
        }
        
        let request = OpenAIRequest {
            model: &self.model,
            messages,
            temperature: self.temperature,
            max_tokens: self.max_tokens,
        };
        
        let response = self.client
            .post(self.provider.url())
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&request)
            .send()
            .map_err(|e| e.to_string())?;
        
        if !response.status().is_success() {
            return Err(format!("API error: {}", response.status()));
        }
        
        let data: OpenAIResponse = response.json().map_err(|e| e.to_string())?;
        
        let text = data.choices.first()
            .map(|c| c.message.content.clone())
            .unwrap_or_default();
        
        let (prompt_tokens, completion_tokens) = data.usage
            .map(|u| (u.prompt_tokens, u.completion_tokens))
            .unwrap_or((0, 0));
        
        let cost = self.calculate_cost(prompt_tokens, completion_tokens);
        
        Ok(LLMResponse {
            text,
            model: self.model.clone(),
            tokens: prompt_tokens + completion_tokens,
            cost,
            latency_ms: 0,
        })
    }
    
    fn chat_anthropic(&self, messages: &[Message]) -> Result<LLMResponse, String> {
        #[derive(Serialize)]
        struct AnthropicRequest<'a> {
            model: &'a str,
            messages: Vec<&'a Message>,
            #[serde(skip_serializing_if = "Option::is_none")]
            system: Option<&'a str>,
            temperature: f32,
            max_tokens: u32,
        }
        
        #[derive(Deserialize)]
        struct AnthropicResponse {
            content: Vec<Content>,
            usage: Option<AnthropicUsage>,
        }
        
        #[derive(Deserialize)]
        struct Content {
            text: String,
        }
        
        #[derive(Deserialize)]
        struct AnthropicUsage {
            input_tokens: u32,
            output_tokens: u32,
        }
        
        let mut system = None;
        let mut user_messages = Vec::new();
        for msg in messages {
            if msg.role == "system" {
                system = Some(msg.content.as_str());
            } else {
                user_messages.push(msg);
            }
        }
        
        let request = AnthropicRequest {
            model: &self.model,
            messages: user_messages,
            system,
            temperature: self.temperature,
            max_tokens: self.max_tokens,
        };
        
        let response = self.client
            .post(self.provider.url())
            .header("x-api-key", &self.api_key)
            .header("Content-Type", "application/json")
            .header("anthropic-version", "2023-06-01")
            .json(&request)
            .send()
            .map_err(|e| e.to_string())?;
        
        if !response.status().is_success() {
            return Err(format!("API error: {}", response.status()));
        }
        
        let data: AnthropicResponse = response.json().map_err(|e| e.to_string())?;
        
        let text = data.content.first()
            .map(|c| c.text.clone())
            .unwrap_or_default();
        
        let (prompt_tokens, completion_tokens) = data.usage
            .map(|u| (u.input_tokens, u.output_tokens))
            .unwrap_or((0, 0));
        
        let cost = self.calculate_cost(prompt_tokens, completion_tokens);
        
        Ok(LLMResponse {
            text,
            model: self.model.clone(),
            tokens: prompt_tokens + completion_tokens,
            cost,
            latency_ms: 0,
        })
    }
    
    fn calculate_cost(&self, prompt_tokens: u32, completion_tokens: u32) -> f64 {
        // Pricing per 1M tokens (input, output)
        let (input_price, output_price) = match self.model.as_str() {
            "anthropic/claude-3-haiku" | "claude-3-haiku-20240307" => (0.25, 1.25),
            "anthropic/claude-3-sonnet" | "claude-3-sonnet-20240229" => (3.0, 15.0),
            "anthropic/claude-3-opus" | "claude-3-opus-20240229" => (15.0, 75.0),
            "openai/gpt-4o" | "gpt-4o" => (5.0, 15.0),
            "openai/gpt-4o-mini" | "gpt-4o-mini" => (0.15, 0.6),
            _ => (0.5, 1.5),
        };
        
        (prompt_tokens as f64 * input_price + completion_tokens as f64 * output_price) / 1_000_000.0
    }
}
